storage/innobase/xtrabackup/src/changed_page_tracking.cc

```diff
@@ -0,0 +1,303 @@
+        "disabled?\n");
+    return (false);
+  } else if (backup_start_lsn > start_lsn) {
+    msg("xtrabackup: pagetracking: Error tracking lsn " LSN_PF
+        " is more than increment "
+        "start lsn " LSN_PF "\n",
+        backup_start_lsn, start_lsn);
+    return (false);
+  }
+
+  /* call component api to generate page tracking file */
+  msg_ts("xtrabackup: pagetracking: calling get pages with start lsn " LSN_PF
+         " and end lsn " LSN_PF "\n",
+         start_lsn, end_lsn);
+  std::ostringstream query;
+  query << "SELECT " << Backup_comp_constants::udf_get_changed_pages << "("
+        << start_lsn << "," << end_lsn << ")";
+
+  char *get_pages_str = read_mysql_one_value(connection, query.str().c_str());
+  auto get_pages = atoi(get_pages_str);
+  free(get_pages_str);
+
+  if (get_pages == 0) {
+    return (true);
+  } else {
+    msg("xtrabackup: pagetracking: Error failed to generate page tracking "
+        "file with error code %d \n",
+        get_pages);
+    return (false);
+  }
+}
+
+/** Set the backupid
+@param[in]   connection  MySQL connection handler
+@param[in]   backupid    current backupid
+return true if able to set */
+bool set_backupid(MYSQL *connection, unsigned long long int backupid) {
+  std::ostringstream query;
+
+  query << "SET GLOBAL " << Backup_comp_constants::mysqlbackup << "."
+        << Backup_comp_constants::backupid << "=\"" << backupid << "\"";
+
+  xb_mysql_query(connection, query.str().c_str(), false);
+
+  query.str("");
+  query.clear();
+  query << "SELECT @@" << Backup_comp_constants::mysqlbackup << "."
+        << Backup_comp_constants::backupid;
+
+  char *component_status_str =
+      read_mysql_one_value(connection, query.str().c_str());
+
+  unsigned long long int get_id = strtoull(component_status_str, NULL, 10);
+  free(component_status_str);
+
+  if (get_id == backupid) {
+    return (true);
+  } else {
+    msg("xtrabackup: pagetracking: Error failed to set backupid"
+        "currently it is set to %llu \n",
+        get_id);
+    return (false);
+  }
+}
+
+/** Check if mysqlbackup component is installed or not
+@param[in]   conn         mysql connection handle
+@return true if installed */
+bool is_component_installed(MYSQL *connection) {
+  char *mysql_component_str =
+      read_mysql_one_value(connection,
+                           "SELECT COUNT(1) FROM mysql.component WHERE "
+                           "component_urn='file://component_mysqlbackup'");
+
+  int mysql_component = strtoull(mysql_component_str, NULL, 10);
+  free(mysql_component_str);
+
+  return mysql_component == 0 ? (false) : (true);
+}
+
+/** Get the last page in the continuous block
+@param[in]    page_set          page_set
+@param[in]    page_no           start page of the block
+@return page id */
+page_no_t range_get_next_page(const xb_page_set page_set, page_no_t page_no) {
+  auto it = page_set.find(page_no);
+  while (true) {
+    auto current_page = *it++;
```

Comment:
isn't this a problem when it is it.end()? it.end()++?

Or is it guaranteed to not return it.end()?

Reply:
we pass valid page_no. even if it last page.  incraase the page id by 1. so the lenght is calculated correctly.